### PR TITLE
Permutation computation refactoring

### PR DESCRIPTION
**Pull request by chrisrichardson**
Generalise face permutation computation to get ready for mixed topology meshes.
* reduce duplicated code in face permutation computation for tet, hex
* adds computation for prism, (should work for pyramid too)

### CODE REVIEW

**Comment by chrisrichardson** [comment]
Would be good to add a test - if anyone can think of a good one @mscroggs @jpdean ?

File: cpp/dolfinx/mesh/permutationcomputation.cpp [hunk 1]
```diff
@@ -21,205 +23,185 @@ using namespace dolfinx;
...
+  // lowest numbered vertex
+  std::int32_t pre = e_vertices[prev[min_v]];
+
+  // post is the (local) number of the next vertex anticlockwise
+  // from the lowest numbered vertex
+  std::int32_t post = e_vertices[prev[3 - min_v]];
+
+  // If min_v is 2 or 3, swap. Why?
+  if (min_v == 2 or min_v == 3)
+    min_v = 5 - min_v;
```
Comment (mscroggs) [inline cpp/dolfinx/mesh/permutationcomputation.cpp:89]: To answer the why: min_v wants to be the number of vertices anticlockwise away the minimum numbered vertex is. If the vertices were numbered:

```
3 -- 2
|    |
0 -- 1
```

We wouldn't need to do the swap, but as we use the numbering:

```
2 -- 3
|    |
0 -- 1
```

we therefore need the swap

Comment (chrisrichardson) [reply cpp/dolfinx/mesh/permutationcomputation.cpp:89]: Thanks @mscroggs . Maybe you can add something in-line to document this.

**Comment by chrisrichardson** [comment]
Fixes issue #3208

**Pull request merged**